cleverhans/future/tf2/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,210 @@
+  :param initial_const: The initial tradeoff-constant to use to tune the relative
+              importance of distance and confidence. If binary_search_steps is large,
+              the initial constant is not important.
+  :param boxmin: Minimum pixel value (default -0.5).
+  :param boxmax: Maximum pixel value (default 0.5).
+  """
+
+  def __init__(self, model, shape=(28, 28, 1), num_labels=10, batch_size=1, confidence=CONFIDENCE, targeted=TARGETED, learning_rate=LEARNING_RATE, binary_search_steps=BINARY_SEARCH_STEPS, max_iterations=MAX_ITERATIONS, abort_early=ABORT_EARLY, initial_const=INITIAL_CONST, boxmin=-0.5, boxmax=0.5):
+    image_size, num_channels = shape[0], shape[2]
+    self.TARGETED = targeted
+    self.LEARNING_RATE = learning_rate
+    self.MAX_ITERATIONS = max_iterations
+    self.BINARY_SEARCH_STEPS = binary_search_steps
+    self.ABORT_EARLY = abort_early
+    self.CONFIDENCE = confidence
+    self.initial_const = initial_const
+    self.batch_size = batch_size
+    self.data_shape = (batch_size,) + shape
+
+    self.repeat = binary_search_steps >= 10
+
+    self.I_KNOW_WHAT_I_AM_DOING_AND_WANT_TO_OVERRIDE_THE_PRESOFTMAX_CHECK = False
+
+    self.shape = (batch_size, image_size, image_size, num_channels)
+    self.boxmul = (boxmax - boxmin) / 2.
+    self.boxplus = (boxmin + boxmax) / 2.
+
+  def attack(self, imgs, targets):
+    """
+    Perform the L_2 attack on the given images for the given targets.
+
+    If self.targeted is true, then the targets represents the target labels.
+    If self.targeted is false, then targets are the original class labels.
+
+    :param imgs: [batch, width, height, channels]
+    :param labs: e.g. if batch is 1, then labs is like [array([0., 0., 0., 0., 0., 0., 1., 0., 0., 0.])]
+    """
+    r = []
+    print('go up to', len(imgs))
+    for i in range(0, len(imgs), self.batch_size):
+      print('tick', i)
+      r.extend(self.attack_batch(imgs[i:i + self.batch_size], targets))
+    return np.array(r)
+
+  def attack_batch(self, imgs, labs):
+    """
+    Run the attack on a batch of images and labels.
+
+    :param imgs: [batch, width, height, channels]
+    :param labs: e.g. if batch is 1, then labs is like [array([0., 0., 0., 0., 0., 0., 1., 0., 0., 0.])]
+    """
+
+    batch_size = self.batch_size
+
+    def compare(x, y):
+      if not isinstance(x, (float, int, np.int64)):
+        x = x.numpy()
+        x = np.copy(x)
+        if self.TARGETED:
+          x[y] -= self.CONFIDENCE
+        else:
+          x[y] += self.CONFIDENCE
+        x = np.argmax(x)
+      if self.TARGETED:
+        return x == y
+      else:
+        return x != y
+
+    # @tf.function # there are some problems to be fixed, open issues in Tensorflow2.0 beta
+    def train_step(modifier, timg, tlab, const):
+      with tf.GradientTape() as tape:
+        newimg = tf.tanh(modifier + timg) * self.boxmul + self.boxplus
+        output = model.predict(newimg)
```

Comment:
the object `model` will not exist here -- it needs to be stashed as an attribute during the `__init__` and referenced as an attribute here (or referenced some other way after a refactor)